core/sup_tree_core/executor_pool/memcache_manager.ex

```diff
@@ -0,0 +1,103 @@
+# Copyright(c) 2015-2018 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule AntikytheraCore.ExecutorPool.MemcacheManager do
+  @moduledoc """
+  A `GenServer` to manage memcache for each executor pool.
+  """
+
+  @timeout         300_000
+  @max_records_num 10
+  @max_record_size 65536
+
+  use GenServer
+  alias AntikytheraCore.Ets.Memcache
+  alias AntikytheraCore.ExecutorPool.RegisteredName, as: RegName
+  alias Antikythera.ExecutorPool.Id, as: EPoolId
+
+  def start_link(name, epool_id) do
+    GenServer.start_link(__MODULE__, epool_id, [name: name])
+  end
+
+  @impl true
+  def init(epool_id) do
+    {:ok, %{epool_id: epool_id, records: %{}, expire_at_set: :gb_sets.new()}, @timeout}
+  end
+
+  @impl true
+  def handle_call({:write, key, value, lifetime, ratio}, _from, state) do
+    evicted_state = evict_expired_records(state, System.monotonic_time(:milliseconds))
+    if :erts_debug.flat_size(value) > @max_record_size do
+      {:reply, {:error, :too_large_object}, evicted_state, @timeout}
+    else
+      new_state =
+        evicted_state
+        |> evict_head_record_if_exceeds_limit()
+        |> do_write(key, value, lifetime, ratio)
+      {:reply, :ok, new_state, @timeout}
+    end
+  end
+
+  @impl true
+  def handle_info(:timeout, state) do
+    new_state = evict_expired_records(state, System.monotonic_time(:milliseconds))
+    {:noreply, new_state, @timeout}
+  end
+
+  defp evict_expired_records(%{records: records, expire_at_set: expire_at_set, epool_id: epool_id} = state, current_time) do
+    if :gb_sets.is_empty(expire_at_set) do
+      state
+    else
+      {{expire_at, key}, new_expire_at_set} = :gb_sets.take_largest(expire_at_set)
+      if expire_at < current_time do
+        Memcache.delete(key, epool_id)
+        new_records = Map.delete(records, key)
+        evict_expired_records(%{state | records: new_records, expire_at_set: new_expire_at_set}, current_time)
+      else
+        state
+      end
+    end
+  end
+
+  defp evict_head_record_if_exceeds_limit(%{records: records, expire_at_set: expire_at_set, epool_id: epool_id} = state) do
+    if map_size(records) == @max_records_num do
```

Comment:
`==` should be `>=` here.
For example when we change `@max_record_num` with hot code upgrade from 10 to 5 and at that time there're 7 cached records, then clients can store as many caches as they want.

Reply:
Oh, I didn't care about the case.
I think we should call `evict_head_record_if_exceeds_limit` recursively if it is the case.

Reply:
Right, let's discard all extra records to keep number of records within the limit.

Reply:
Fixed in 6ae7fc15a013142f6442c519f27080b81a63bbc9